src/Bedrock.Framework.Experimental/Protocols/Kafka/KafkaClientConnectionContext.cs

```diff
@@ -0,0 +1,131 @@
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Connections;
+using Bedrock.Framework.Protocols.Kafka.Messaging;
+using Bedrock.Framework.Protocols.Kafka.Utils;
+
+namespace Bedrock.Framework.Protocols.Kafka
+{
+    /// <summary>
+    /// Encapsulates a client Kafka connection that channels requests to an underlying connection
+    /// in the order that they were written; and reads responses in the same order.
+    /// </summary>
+    public class KafkaClientConnectionContext : IKafkaClientConnectionContext
+    {
+        private readonly Channel<(object, IKafkaMessageDescriptor, TaskCompletionSource<object>)> _requestChannel =
+            Channel.CreateUnbounded<(object, IKafkaMessageDescriptor, TaskCompletionSource<object>)>(
+                new UnboundedChannelOptions()
+                {
+                    SingleReader = true,
+                    SingleWriter = false
+                });
+
+        private readonly ConcurrentQueue<(int, IKafkaMessageDescriptor, TaskCompletionSource<object>)> _requestQueue =
+            new ConcurrentQueue<(int, IKafkaMessageDescriptor, TaskCompletionSource<object>)>();
+
+        private readonly IMessageWriter<KafkaRequest> _requestWriter;
+        private readonly IMessageReader<KafkaResponse> _responseReader;
+
+        private int _correlationId = 0;
+
+        /// <summary>
+        /// Initializes a new <see cref="KafkaClientConnectionContext"/>.
+        /// </summary>
+        /// <param name="requestWriter">An <see cref="IMessageWriter{KafkaRequest}"/> that can write Kafka requests to
+        /// a buffer.</param>
+        /// <param name="responseReader">An <see cref="IMessageReader{KafkaResponse}"/> that can read Kafka responses from
+        /// a buffer.</param>
+        public KafkaClientConnectionContext(IMessageWriter<KafkaRequest> requestWriter, IMessageReader<KafkaResponse> responseReader)
+        {
+            _requestWriter = requestWriter;
+            _responseReader = responseReader;
+        }
+
+        /// <inheritdoc/>
+        public Task StartAsync(ConnectionContext connectionContext, CancellationToken cancellationToken = default)
+        {
+            _ = WriteRequestsAsync(connectionContext, cancellationToken);
+            _ = ReadResponsesAsync(connectionContext, cancellationToken);
+
+            return Task.CompletedTask;
+        }
+
+        /// <inheritdoc/>
+        public async Task<TResponseSchema> SendAsync<TRequestSchema, TResponseSchema>(
+            TRequestSchema request,
+            IKafkaMessageDescriptor messageDescriptor,
+            CancellationToken cancellationToken = default)
+        {
+            var tcs = new TaskCompletionSource<object>();
+
+            _requestChannel.Writer.TryWrite((request, messageDescriptor, tcs));
+
+            return (TResponseSchema)await tcs.Task;
+        }
+
+        private async Task WriteRequestsAsync(ConnectionContext connectionContext, CancellationToken cancellationToken = default)
```

Comment:
hmm, might steal some of this, but...
This only does a single request-response on the wire at any one time. 
"i.e., clients can send requests even while awaiting responses for preceding requests" (from https://kafka.apache.org/protocol#protocol_network)
That's what I'm tryin to do with the ValueTaskSource shenanigans at https://github.com/AndyPook/A6k.Kafka/blob/ba28b5d91f306dddde9872e4a66854adb3d2ad5f/A6k.Kafka/KafkaConnection.cs#L149

Not really a problem at this point though

Reply:
So, the `KafkaClientConnectionContext` just buffers the writes through the channel to ensure that their order is preserved. The protocol (`KafkaClientProtocol`) sat above could quite easily call `SendAsync()` several times which would result in several inflight (and ordered) requests which complete as their responses arrive.

Reply:
🤔 they'd be in-flight in so far as they are "queued" on the Channel. But there wouldn't be multiple requests on the wire at once. Kafka suggests that clients should be able to physically send multiple requests on the wire. So you can send request 2 before the response for request 1 comes back.
If I'm reading this right (??) doesn't this loop mean that response 1 must comes back before request 2 is sent.

Obviously this is just an optimization. Really not important for right now. Can be replaced later

Reply:
`SendAsync()` writes to the request channel and then waits for the `TaskCompletionSource` to complete. The request channel can have many writers. The single reader (implemented in the `WriteRequestsAsync()` method) will read from the request channel and forward it onto the underlying `ConnectionContext`. It will keep doing that until there are no more requests on the request channel in which case it will block until more arrive.

The `ReadResponsesAsync()` method reads from the underlying `ConnectionContext` and then dequeues the request queue. That item represents the request for the response that we have just received, allowing us to complete the associated Task that is being awaited upon.

The two loops around the underlying `ConnectionContext` will just keep pushing and pulling as data arrives. As such you can send many requests while you're still waiting for a response, but, as per the protocol, when you do get a response, it should be for the first request and their subsequent order thereafter.

Reply:
Doh! completely miss-read it.
Carry on, nothing to see here

Reply:
right... I guess the only real diff to "my" approach is:
 * tuple vs my Op object
 * TCS vs my attempt at ValueTaskSource (in the Op type)

Is that how you'd see it?